tests/rptest/tests/tombstone_removal_e2e_test.py

```diff
@@ -0,0 +1,283 @@
+            producer.free()
+
+    def leadership_transferred_to_node(self, node_id: int) -> bool:
+        current_leader = self.admin.get_partition("kafka", self.topic_name, 0)[
+            "leader_id"
+        ]
+        if current_leader == node_id:
+            return True
+        self.redpanda.logger.debug(
+            f"Current leader is {current_leader}, trying to transfer to node {node_id}"
+        )
+        try:
+            self.admin.transfer_leadership_to(
+                namespace="kafka",
+                topic=self.topic_name,
+                partition=0,
+                target_id=node_id,
+            )
+        except requests.exceptions.HTTPError as e:
+            self.redpanda.logger.debug(
+                f"Failed to transfer leadership to node {node_id}, retrying: {e}"
+            )
+        return False
+
+    def transfer_leadership_to_node(self, node_id: int):
+        wait_until(
+            lambda: self.leadership_transferred_to_node(node_id),
+            timeout_sec=30,
+            backoff_sec=1,
+            err_msg=f"timed out waiting for leadership transfer to node {node_id}",
+        )
+
+    def get_data59_from_node(self, node_id: int):
+        self.transfer_leadership_to_node(node_id)
+
+        consumed = self.rpk.consume(
+            topic=self.topic_name,
+            partition=0,
+            format="%k=%V at offset %o\n",
+            n=100,
+            offset=":end",
+            timeout=100,  # seconds
+        )
+        lines = consumed.strip().split("\n")
+        data59 = {
+            line
+            for line in lines
+            if "key-5=" <= line[0:6] <= "key-9="
+            and not line.split("=")[1].startswith("0 at offset")
+        }
+        self.redpanda.logger.debug(
+            f"Consumed lines for keys 5-9 from {node_id=}: {data59}"
+        )
+        return data59
+
+    # make sure it removed all data batches on the leader
+    def no_more_than_n_records_below_offset(self, n: int, offset: int) -> bool:
+        offsets = [
+            int(x)
+            for x in self.rpk.consume(
+                topic=self.topic_name,
+                n=n + 1,
+                partition=0,
+                format="%o ",
+                offset=f":{offset}",
+                timeout=5,  # seconds
+            )
+            .strip()
+            .split(" ")
+        ]
+        assert offsets is not None, "rpk consume returned None"
+        self.redpanda.logger.debug(
+            f"{n + 1=} offsets below {offset=}: {offsets}, checking <= {n=}"
+        )
+        self.redpanda.logger.debug(
+            f"offsets below {offset=}: {[1 for o in offsets if o < offset]}"
+            f", sum={sum(1 for o in offsets if o < offset)}, {n=}"
+        )
+        # rpk consume doesn't seem to honour offsets exactly
+        return sum(1 for o in offsets if o < offset) <= n
+
+    @cluster(num_nodes=6)  # 5 for cluster + 1 for producer
+    def test_single_slow_follower(self):
```

Comment:
please add some prose code doc to describe the scenario

helpful for readers, helpful for test results.html to quickly know what the test is doing and triage failures for someone less familiar with the test